File: libraries/ads/src/commonMain/kotlin/com/dangerfield/drop2048/libraries/ads/AdNetwork.kt
@file:OptIn(ExperimentalObjCName::class)

package com.dangerfield.drop2048.libraries.ads

import me.tatarka.inject.annotations.Inject
import software.amazon.lastmile.kotlin.inject.anvil.AppScope
import software.amazon.lastmile.kotlin.inject.anvil.ContributesBinding
import software.amazon.lastmile.kotlin.inject.anvil.SingleIn
import kotlin.experimental.ExperimentalObjCName
import kotlin.native.ObjCName

/**
 * The three SDK formats the app serves.
 *
 * [Banner] arrived on 2026-09-20 and overturned SPEC 12.4's "Banners: none",
 * which this enum used to state in prose. The reasoning that rule was built on
 * has not been thrown away, it has been answered: the banner is drawn **only
 * where the arrow row would have been**, so it never takes vertical space the
 * board was using, and it occupies no space at all until an ad is really on
 * screen, so a build with no fill is a build with no strip. D28 has the full
 * trade, including the principle that was spent to get it.
 *
 * [Banner] is also the odd one out mechanically: [AdNetwork.show] cannot serve
 * it, because a banner is a view that lives in the layout rather than a
 * full-screen thing that is shown and dismissed. It is served by
 * [BannerSurface], and the entry here exists so [AdUnits] has somewhere to keep
 * the unit id and so a future network cannot forget the format exists.
 */
@ObjCName("AdFormat", exact = true)
enum class AdFormat {
    Rewarded,
    Interstitial,
    Banner,
}

/**
 * What the SDK did, flattened to something an ObjC-facing enum can carry.
 *
 * Deliberately *not* [RewardOutcome]: the Swift side has to be able to build a
 * return value, and a sealed hierarchy of Kotlin data objects is awkward to
 * construct from Swift. `RealAdGate` widens this back into the richer sealed
 * type, which is also the one place the fail-open mapping lives.
 */
@ObjCName("AdShowResult", exact = true)
enum class AdShowResult {
    /** Watched to the reward threshold. */
    Rewarded,

    /** Closed early, by the player, on purpose. The only result that withholds. */
    Dismissed,

    /** The network had nothing to serve. */
    NoFill,

    /** The SDK reported no route to the network. */
    Offline,

    /** Nothing was attempted — the SDK is not initialised, or has no ad loaded. */
    NotShown,

    /** Anything else. [AdShowOutcome.errorKind] carries the detail, for telemetry only. */
    Failed,
}

/**
 * A result plus an optional machine-readable failure kind. Never a message for
 * the player — an ad failure is invisible to them by design.
 */
@ObjCName("AdShowOutcome", exact = true)
class AdShowOutcome(
    val result: AdShowResult,
    val errorKind: String? = null,
)

/**
 * The thin platform seam under [AdGate] and [InterstitialGate]: load an ad, show
 * it, say what happened.
 *
 * Everything policy-shaped — every gate in SPEC 12.3, the caps, the rule that a
 * failed rewarded ad still pays the player — sits above this in common code, so
 * it is the same on both platforms and testable without an ad network.
 *
 * **Android binding** is `AdMobAdNetwork` in `:libraries:ads:impl/androidMain`.
 * **iOS has no real binding yet** and falls through to [NotWiredAdNetwork].
 *
 * Implementations must not throw. An ad SDK that blows up has to look like
 * [AdShowResult.Failed] from here.
 */
@ObjCName("AdNetwork", exact = true)
interface AdNetwork {

    /**
     * Consent first, then SDK init. On Android that is UMP then
     * `MobileAds.initialize`; on iOS it will be UMP **then** ATT **then**
     * `MobileAds.start`. Both orders are a store requirement rather than a
     * preference — an ad request that beats the consent form is a policy
     * violation — so every show path awaits this before it touches the SDK.
     *
     * Idempotent, and cheap after the first call.
     */
    suspend fun prepare()

    /**
     * Shows an already-loaded ad, or loads one first. Suspends until it closes.
     * The implementation resolves the ad unit itself from [AdUnits], so there is
     * exactly one file to edit when the real ids arrive.
     *
     * [AdFormat.Banner] has no answer here and must return
     * [AdShowResult.NotShown]. A banner is a view in a layout, not something
     * that is shown over the app and dismissed, so there is nothing for this
     * call to suspend on. [BannerSurface] serves it.
     */
    suspend fun show(format: AdFormat): AdShowOutcome

    /**
     * Whether a [format] is loaded and showable **right now**, without a network
     * round trip.
     *
     * SPEC 12.3's "preloaded, and skipped silently if not ready" is the only
     * reason this exists. Without it the interstitial path has to attempt a show
     * to discover there is nothing to show, and an attempt is a wait.
     */
    fun isReady(format: AdFormat): Boolean

    /** Warms a format. Fire and forget; failures are the SDK's problem, not the caller's. */
    fun preload(format: AdFormat)
}

/**
 * The network on a platform with no SDK wired, and the single most important
 * thing in this file is what it does **not** return.
 *
 * It never answers [AdShowResult.Rewarded]. Sodogku shipped an iOS ad stub that
 * paid out rewards without showing anything, and its own post-mortem is the
 * reason this class is written down rather than defaulted: a reward that grants
 * without an ad is worse than no ads at all, because it is invisible. Every log
 * line, every dashboard and every test reads exactly as it would if the ad had
 * played.
 *
 * [AdShowResult.NotShown] with a named [AdShowOutcome.errorKind] is the honest
 * answer, and it is the one the layer above already knows how to handle: the
 * rewarded path pays the player anyway because that is a deliberate rule applied
 * to a real outcome, and the interstitial path shows nothing because there is
 * nothing to show. Neither is a lie about an impression.
 *
 * iOS needs `IOSAdNetwork` in Swift plus the GoogleMobileAds package on the
 * Xcode target; both are on `OWNER-TODO.md`.
 */
@SingleIn(AppScope::class)
@ContributesBinding(AppScope::class)
@Inject
class NotWiredAdNetwork : AdNetwork {
    override suspend fun prepare() = Unit

    override suspend fun show(format: AdFormat): AdShowOutcome =
        AdShowOutcome(AdShowResult.NotShown, errorKind = "ad_network_not_wired")

    override fun isReady(format: AdFormat): Boolean = false

    override fun preload(format: AdFormat) = Unit
}

File: apps/ios/iosApp/Platform/AdNetwork.swift
//
//  AdNetwork.swift
//  iosApp
//
//  AdMob on iOS, plus the two consent prompts that must run in front of it.
//
//  ORDER IS A POLICY REQUIREMENT, NOT A PREFERENCE:
//    1. UMP consent (EEA / UK)
//    2. App Tracking Transparency
//    3. MobileAds.start()
//    4. the first ad request
//
//  Google's own guidance is that ATT comes after the UMP form, because UMP can
//  present the ATT pre-prompt. Apple rejects builds that request the IDFA
//  before the ATT dialog. Neither failure shows up as a crash; both show up as
//  a rejected release, which is why the order lives in one place.
//
//  Everything policy-shaped above this — frequency gates, the new-user grace,
//  the offline grace, and the rule that a failed ad still pays the player —
//  is in `RealAdGate` in shared Kotlin. This file only loads and shows.
//
//  The `#if canImport(GoogleMobileAds)` guards are kept because the whole point
//  of them is that a missing SDK degrades to "every ad fails", which the shared
//  Kotlin already pays the player for. They are no longer the live path: the
//  `GoogleMobileAds` product (which carries `UserMessagingPlatform` with it) is
//  a package dependency of the iosApp target, so both branches are real and the
//  `#else` is the thing that should never fire.
//
//  The app id lives in `Info.plist` under `GADApplicationIdentifier` and the
//  unit ids in `AdUnits.kt`. Both are Google's published test values today;
//  they flip together when `AdUnits.useTestUnits` does.
//
//  `AdUnits` resolves here only because `:libraries:ads` is exported to the
//  framework: `ApplicationConventionPlugin` has the `export(...)` and
//  `apps/compose/build.gradle.kts` makes it an `api` dependency, because
//  Kotlin/Native refuses to export a dependency that is not one. Without both,
//  `AdUnits.shared.ios(format:)` does not compile — nothing in Kotlin reads
//  `AdUnits` on iOS, so it would not otherwise be reachable from the
//  framework's API.
//
//  Do not answer a resolution failure by hardcoding the unit id here.
//  `AdUnits.kt` is deliberately the only file that carries one, and the failure
//  it guards against is exactly a half-migrated app with a real id in Kotlin
//  and a test id in Swift.
//
import ComposeApp
import Foundation
import UIKit

#if canImport(AppTrackingTransparency)
import AppTrackingTransparency
#endif

#if canImport(GoogleMobileAds)
import GoogleMobileAds
#endif

#if canImport(UserMessagingPlatform)
import UserMessagingPlatform
#endif

class IOSAdNetwork: NSObject, AdNetwork {

    private var initialised = false

    // MARK: - AdNetwork

    func __prepare() async throws {
        guard !initialised else { return }
        #if canImport(GoogleMobileAds)
        await requestConsent()
        await requestTrackingAuthorization()
        guard canRequestAds() else { return }
        // The app is general-audience, matching Android's
        // `TAG_FOR_CHILD_DIRECTED_TREATMENT_FALSE`. Set before `start()`, as on
        // Android, so no request can go out ahead of the configuration.
        //
        // `.unspecified` is this SDK's current spelling of that: the
        // age-treatment enum replaced the child-directed and
        // under-age-of-consent flags, and it has no "definitely not restricted"
        // case — declaring neither child nor teen *is* the general-audience
        // answer. It is also the default, so this line is a statement of the
        // decision rather than a change of behaviour.
        MobileAds.shared.requestConfiguration.ageRestrictedTreatment = .unspecified
        _ = await MobileAds.shared.start()
        initialised = true
        #endif
    }

    // `AdFormat` is qualified because GoogleMobileAds exports a type of that
    // name too (`GADAdFormat`), and an unqualified mention is ambiguous once
    // both modules are imported. The Kotlin one is the contract being
    // implemented here.
    func __show(format: ComposeApp.AdFormat) async throws -> AdShowOutcome {
        // A banner is a view in the game's own layout, served by
        // `BannerSurface`, so there is nothing here to show and nothing to
        // suspend on. Answered before `prepare()` so it costs no consent
        // round trip. Mirrors `AdMobAdNetwork`'s `banner_is_not_shown`.
        if format == .banner {
            return AdShowOutcome(result: .notShown, errorKind: "banner_is_not_shown")
        }
        try await __prepare()
        #if canImport(GoogleMobileAds)
        guard initialised else {
            return AdShowOutcome(result: .notShown, errorKind: "sdk_not_ready")
        }
        guard let root = await Self.rootViewController() else {
            return AdShowOutcome(result: .notShown, errorKind: "no_root_view_controller")
        }
        switch format {
        case .rewarded:
            return await showRewarded(from: root)
        case .interstitial:
            return await showInterstitial(from: root)
        default:
            return AdShowOutcome(result: .notShown, errorKind: "banner_is_not_shown")
        }
        #else
        return AdShowOutcome(result: .notShown, errorKind: "google_mobile_ads_not_linked")
        #endif
    }

    /// Android answers `false` for every banner here too: a banner's readiness
    /// is the `AdView`'s own business, not this seam's.
    func isReady(format: ComposeApp.AdFormat) -> Bool {
        #if canImport(GoogleMobileAds)
        switch format {
        case .rewarded: return cachedRewarded != nil
        case .interstitial: return cachedInterstitial != nil
        default: return false
        }
        #else
        return false
        #endif
    }

    func preload(format: ComposeApp.AdFormat) {
        #if canImport(GoogleMobileAds)
        guard format != .banner else { return }
        Task { [weak self] in
            try? await self?.__prepare()
            guard self?.initialised == true else { return }
            switch format {
            case .rewarded:
                guard self?.cachedRewarded == nil else { return }
                self?.cachedRewarded = try? await RewardedAd.load(
                    with: AdUnits.shared.ios(format: .rewarded), request: Request())
            case .interstitial:
                guard self?.cachedInterstitial == nil else { return }
                self?.cachedInterstitial = try? await InterstitialAd.load(
                    with: AdUnits.shared.ios(format: .interstitial), request: Request())
            default:
                return
            }
        }
        #endif
    }

    #if canImport(GoogleMobileAds)

    private var cachedRewarded: RewardedAd?
    private var cachedInterstitial: InterstitialAd?

    /// No reward and no reward handler: closing is the ordinary ending, and
    /// `.dismissed` is how the shared gate hears "it was on screen".
    private func showInterstitial(from root: UIViewController) async -> AdShowOutcome {
        let ad: InterstitialAd
        do {
            if let cached = cachedInterstitial {
                ad = cached
                cachedInterstitial = nil
            } else {
                ad = try await InterstitialAd.load(
                    with: AdUnits.shared.ios(format: .interstitial), request: Request())
            }
        } catch {
            return Self.loadFailure(error)
        }

        let delegate = DismissalDelegate()
        ad.fullScreenContentDelegate = delegate
        let dismissal = await withCheckedContinuation { (continuation: CheckedContinuation<Dismissal, Never>) in
            delegate.onFinished = { continuation.resume(returning: $0) }
            Task { @MainActor in
                ad.present(from: root)
            }
        }

        switch dismissal {
        case .failed(let kind): return AdShowOutcome(result: .failed, errorKind: kind)
        case .closed: return AdShowOutcome(result: .dismissed, errorKind: nil)
        }
    }

    private func showRewarded(from root: UIViewController) async -> AdShowOutcome {
        let ad: RewardedAd
        do {
            if let cached = cachedRewarded {
                ad = cached
                cachedRewarded = nil
            } else {
                ad = try await RewardedAd.load(
                    with: AdUnits.shared.ios(format: .rewarded), request: Request())
            }
        } catch {
            return Self.loadFailure(error)
        }

        let delegate = DismissalDelegate()
        ad.fullScreenContentDelegate = delegate
        var earned = false
        let dismissal = await withCheckedContinuation { (continuation: CheckedContinuation<Dismissal, Never>) in
            delegate.onFinished = { continuation.resume(returning: $0) }
            Task { @MainActor in
                ad.present(from: root) { earned = true }
            }
        }
        preload(format: .rewarded)

        switch dismissal {
        case .failed(let kind): return AdShowOutcome(result: .failed, errorKind: kind)
        case .closed: return AdShowOutcome(result: earned ? .rewarded : .dismissed, errorKind: nil)
        }
    }

    /// Main-actor because the second half presents a view controller. Both
    /// failures are swallowed on purpose: a consent update that errors leaves
    /// `canRequestAds` false, which `prepare` already reads as "no ads today".
    @MainActor
    private func requestConsent() async {
        try? await ConsentInformation.shared.requestConsentInfoUpdate(with: RequestParameters())
        guard let root = Self.rootViewController() else { return }
        try? await ConsentForm.loadAndPresentIfRequired(from: root)
    }

    /// `canRequestAds` is the gate, not "did the form show". UMP answers true
    /// for a user outside the EEA who was never shown anything, and false for
    /// one who declined; reading the form's presence would block ads for most
    /// of the world.
    private func canRequestAds() -> Bool {
        ConsentInformation.shared.canRequestAds
    }

    private static func loadFailure(_ error: Error) -> AdShowOutcome {
        let code = (error as NSError).code
        let kind = "load_\(code)"
        // Qualified: UserMessagingPlatform exports a `RequestError` of its own.
        switch GoogleMobileAds.RequestError.Code(rawValue: code) {
        // A timed-out load is Android's `load_timeout`, and it says the same
        // thing about the player's situation: nothing to watch right now.
        case .noFill, .timeout: return AdShowOutcome(result: .noFill, errorKind: kind)
        case .networkError: return AdShowOutcome(result: .offline, errorKind: kind)
        default: return AdShowOutcome(result: .failed, errorKind: kind)
        }
    }

    private enum Dismissal {
        case closed
        case failed(String)
    }

    private final class DismissalDelegate: NSObject, FullScreenContentDelegate {
        var onFinished: ((Dismissal) -> Void)?

        func adDidDismissFullScreenContent(_ ad: FullScreenPresentingAd) {
            finish(.closed)
        }

        func ad(_ ad: FullScreenPresentingAd, didFailToPresentFullScreenContentWithError error: Error) {
            finish(.failed("show_\((error as NSError).code)"))
        }

        private func finish(_ dismissal: Dismissal) {
            let handler = onFinished
            onFinished = nil
            handler?(dismissal)
        }
    }

    #endif

    /// ATT goes in front of the first ad request, and it is asked at a moment
    /// where the value is legible — which is why `prepare()` is called lazily by
    /// the first ad gate rather than at launch.
    private func requestTrackingAuthorization() async {
        #if canImport(AppTrackingTransparency)
        guard #available(iOS 14.5, *) else { return }
        guard ATTrackingManager.trackingAuthorizationStatus == .notDetermined else { return }
        _ = await ATTrackingManager.requestTrackingAuthorization()
        #endif
    }

    @MainActor
    private static func rootViewController() -> UIViewController? {
        UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .first { $0.activationState == .foregroundActive }?
            .keyWindow?
            .rootViewController
    }
}
